src/nupic/utils/SlidingWindow.hpp

```diff
@@ -0,0 +1,163 @@
+/* Numenta Platform for Intelligent Computing (NuPIC)
+ * Copyright (C) 2016, Numenta, Inc.  Unless you have an agreement
+ * with Numenta, Inc., for a separate license for this software code, the
+ * following terms and conditions apply:
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU Affero Public License version 3 as
+ * published by the Free Software Foundation.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
+ * See the GNU Affero Public License for more details.
+ *
+ * You should have received a copy of the GNU Affero Public License
+ * along with this program.  If not, see http://www.gnu.org/licenses.
+ *
+ * http://numenta.org/licenses/
+ * ---------------------------------------------------------------------
+ */
+
+
+#ifndef NUPIC_UTIL_SLIDING_WINDOW_HPP
+#define NUPIC_UTIL_SLIDING_WINDOW_HPP
+
+#include <vector>
+
+#include <nupic/types/Types.hpp>
+#include <nupic/utils/Log.hpp>
+
+namespace nupic {
+  namespace util {
+
+    template<class T> 
+    class SlidingWindow {
+      public:
+        SlidingWindow(UInt maxCapacity);
+        SlidingWindow(UInt maxCapacity, std::vector<T> initialData);
```

Comment:
Don't pass a `std::vector<T>` by value. Do a `const std::vector<T>&`.

Even better, you could use iterators, especially since this is already a templated class so you have to implement it in the header. Iterators maximize reusability -- your current code requires the caller to put their data in a vector, so for example, if the data is in a numpy array it will need to be copied into a vector.

~~~c++
template<typename Iterator>
SlidingWindow(UInt maxCapacity, Iterator initialData_begin, Iterator initialData_end);
~~~